### PR TITLE
add checked methods for iterator

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

Only for tests, don't merge.

### CODE REVIEW

File: src/rocksdb.rs [hunk 1]
```diff
@@ -250,17 +293,27 @@ impl<D> DBIterator<D> {
...
+        let valid = self.checked_valid().expect("checked_valid fail");
+        if valid {
             Some((self.key().to_vec(), self.value().to_vec()))
         } else {
             None
         }
     }
 
+    #[deprecated]
     pub fn valid(&self) -> bool {
```
Comment (breeswish) [inline src/rocksdb.rs]: Maybe better to keep the `valid()` interface to be aligned with the RocksDB C/C++ API.

File: src/rocksdb.rs [hunk 2]
```diff
@@ -250,17 +293,27 @@ impl<D> DBIterator<D> {
...
         unsafe { crocksdb_ffi::crocksdb_iter_valid(self.inner) }
     }
 
+    pub fn checked_valid(&self) -> Result<bool, String> {
```
Comment (breeswish) [inline src/rocksdb.rs]: How about call it `valid_status()`?

File: src/rocksdb.rs [hunk 3]
```diff
@@ -240,7 +283,7 @@ impl<D> DBIterator<D> {
     }
 
     pub fn value(&self) -> &[u8] {
-        assert!(self.valid());
+        assert_eq!(self.checked_valid(), Ok(true));
```
Comment (breeswish) [inline src/rocksdb.rs]: If the original `valid()` is preserved, we can still use `valid()` here.

File: src/rocksdb.rs [hunk 4]
```diff
@@ -250,17 +293,27 @@ impl<D> DBIterator<D> {
     }
 
     pub fn kv(&self) -> Option<(Vec<u8>, Vec<u8>)> {
```
Comment (breeswish) [inline src/rocksdb.rs]: Maybe we should pass the `Result` out

**Comment by yiwu-arbug** [comment]
I think we don't have API compatibility promise for rust-rocskdb? In that case I prefer simply change existing method signature.

**Comment by zhangjinpeng1987** [comment]
Seems you use the wrong branch? I remember @yiwu-arbug has changed tikv-3.0 & tikv-3.1 to use the branch of tikv-3.x branch of rust-rocksdb.

**Comment by yiwu-arbug** [comment]
yup, we've change to use `tikv-3.x` branch instead of `tikv-3.0` since both 3.0 and 3.1 are using the same rust-rocksdb branch.

**Pull request closed without merging**